sktime/decomposition/vampnet.py

```diff
@@ -0,0 +1,662 @@
+        # Calculate eigvalues and eigvectors
+        identity = torch.eye(mat.shape[0], dtype=mat.dtype, device=mat.device)
+        eigval, eigvec_t = torch.symeig(mat + epsilon*identity, eigenvectors=True)
+        eigval = torch.abs(eigval)
+        eigvec = eigvec_t.transpose(0, 1)
+    elif mode == 'clamp':
+        eigval, eigvec_t = torch.symeig(mat, eigenvectors=True)
+        eigval = torch.clamp_min(eigval, min=epsilon)
+        eigvec = eigvec_t.transpose(0, 1)
+    else:
+        raise RuntimeError("invalid mode!")
+
+    # Build the diagonal matrix with the filtered eigenvalues or square
+    # root of the filtered eigenvalues according to the parameter
+    if ret_sqrt:
+        diag = torch.diag(torch.sqrt(1. / eigval))
+    else:
+        diag = torch.diag(1. / eigval)
+
+    return torch.chain_matmul(eigvec_t, diag, eigvec)
+
+
+sym_inverse.valid_modes = ('trunc', 'regularize', 'clamp')
+
+
+def koopman_matrix(x: torch.Tensor, y: torch.Tensor, epsilon: float = 1e-6, mode: str = 'trunc',
+                   c_xx: Optional[Tuple[torch.Tensor, torch.Tensor, torch.Tensor]] = None) -> torch.Tensor:
+    r""" Computes the Koopman matrix
+
+    .. math:: K = C_{00}^{-1/2}C_{0t}C_{tt}^{-1/2}
+
+    based on data over which the covariance matrices :math:`C_{\cdot\cdot}` are computed.
+
+    Parameters
+    ----------
+    x : torch.Tensor
+        Instantaneous data.
+    y : torch.Tensor
+        Time-lagged data.
+    epsilon : float, default=1e-6
+        Cutoff parameter for small eigenvalues.
+    mode : str, default='trunc'
+        Regularization mode for Hermetian inverse. See :meth:`sym_inverse`.
+    c_xx : tuple of torch.Tensor, optional, default=None
+        Tuple containing c00, c0t, ctt if already computed.
+
+    Returns
+    -------
+    K : torch.Tensor
+        The Koopman matrix.
+    """
+    if c_xx is not None:
+        c00, c0t, ctt = c_xx
+    else:
+        c00, c0t, ctt = covariances(x, y, remove_mean=True)
+    c00_sqrt_inv = sym_inverse(c00, ret_sqrt=True, epsilon=epsilon, mode=mode)
+    ctt_sqrt_inv = sym_inverse(ctt, ret_sqrt=True, epsilon=epsilon, mode=mode)
+    return torch.chain_matmul(c00_sqrt_inv, c0t, ctt_sqrt_inv).t()
+
+
+def covariances(x: torch.Tensor, y: torch.Tensor, remove_mean: bool = True):
+    """Computes instantaneous and time-lagged covariances matrices.
+
+    Parameters
+    ----------
+    x : (T, n) torch.Tensor
+        Instantaneous data.
+    y : (T, n) torch.Tensor
+        Time-lagged data.
+    remove_mean: bool, default=True
+        Whether to remove the mean of x and y.
+
+    Returns
+    -------
+    cov_00 : (n, n) torch.Tensor
+        Auto-covariance matrix of x.
+    cov_0t : (n, n) torch.Tensor
+        Cross-covariance matrix of x and y.
+    cov_tt : (n, n) torch.Tensor
+        Auto-covariance matrix of y.
+
+    See Also
+    --------
+    sktime.covariance.Covariance : Estimator yielding these kind of covariance matrices based on raw data.
```

Comment:
these kind of covariance matrices? raw data vs what?

Reply:
vs tensors in an autograd framework

Reply:
I agree though, could be phrased better.